text/0000-default-field-values.md

```diff
@@ -0,0 +1,1807 @@
+provide implementations of `Default`. Rather, the macro facilitates the
+generation of `MyType::new` constructors.
+
+For example, you may write:
+
+```rust
+#[derive(new)]
+struct Foo {
+    x: bool,
+    #[new(value = "42")]
+    y: i32,
+    #[new(default)]
+    z: Vec<String>,
+}
+
+Foo::new(true);
+
+#[derive(new)]
+enum Enum {
+    FirstVariant,
+    SecondVariant(bool, #[new(default)] u8),
+    ThirdVariant { x: i32, #[new(value = "vec![1]")] y: Vec<u8> }
+}
+
+Enum::new_first_variant();
+Enum::new_second_variant(true);
+Enum::new_third_variant(42);
+```
+
+Notice how `#[new(value = "vec![1]")`, `#[new(value = "42")]`,
+and `#[new(default)]` are used to provide values that are then omitted
+from the respective constructor functions that are generated.
+
+If you transcribe the above snippet as much as possible to the system proposed
+in this RFC, you would get:
+
+```rust
+struct Foo {
+    x: bool,
+    y: i32 = 42,
+    z: Vec<String> = <_>::default(),
+    //               --------------
+    //               note: assuming some `impl const Default { .. }` mechanism.
+}
+
+Foo { x: true };
+
+enum Enum {
+    FirstVariant,
+    SecondVariant(bool, u8), // See future possibilities.
+    ThirdVariant { x: i32, y: Vec<u8> = vec![1] }
+}
+
+Enum::FirstVariant;
+Enum::SecondVariant(true, 0);
+Enum::ThirdVariant { x: 42 };
+```
+
+Relative to `#[derive(new)]`, the main benefits are:
+
++ No wrapping code in strings, as noted in previous sections.
++ The defaults used can be mixed and matches; it works to request all defaults
+  or just some of them.
+
+The constructor functions `new_first_variant(..)` are not provided for you.
+However, it should be possible to tweak `#[derive(new)]` to interact with
+this RFC so that constructor functions are regained if so desired.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. What is the right interaction wrt. `#[non_exhaustive]`?
+
+   In particular, if given the following definition:
+
+   ```rust
+   #[non_exhaustive]
+   pub struct Config {
+       pub height: u32,
+       pub width: u32,
+   }
+   ```
+
+   it could be possible to construct a `Config` like so, if the construction of
+   types without default field values is allowed (to support semver changes):
+
+   ```rust
+   let config = Config { width: 640, height: 480, .. };
+   ```
+
+   then adding a field to `Config` can only happen iff that field
+   is provided a default value.
+
+   This arrangement, while diminishing the usefulness of `#[non_exhaustive]`,
+   makes the ruleset of the language simpler, more consistent, and also
+   simplifies type checking as `#[non_exhaustive]` is entirely ignored
+   when checking `Foo { fields, .. }` expressions.
+
+   As an alternative, users who desire the semantics described above can
+   omit `#[non_exhaustive]` from their type and instead add a private
+   defaulted field that has a ZST.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
```

Comment:
If we want to preserve the ability to use this syntax in const, `#[derive(const Default)]` could be an interesting way of still getting the shorthand I seem to care so much about.

Reply:
To be more specific, what would you think of allowing this, assuming we had such a thing as `const Trait` bounds?

```rust
#[derive(const Default)]
struct State {
    foo: HashMap<u32, String>,
    bar: Arc<Mutex<Vec<Bar>>>,
    baz: i32 = 5,
}

let state = State { .. };
```

This seems nicer than making the struct author write `= Default::default()` many times. Structs which do not wish to provide a default for all fields would, of course, not implement Default (const or otherwise).

Attempting to use the `..` syntax with a struct that implements non-const `Default` would result in an error message saying the struct needs to implement `const Default` and suggesting to use `..Struct::default()` (if no non-accessible fields), field-wise initialization otherwise.

Reply:
I have some concerns around this syntax.

The proposed code is equivalent to the following, right?

```rust
const fn d<D: Default>() -> D {
    D::default()
}
#[derive(Default)]
struct State {
    foo: HashMap<u32, String> = d(),
    bar: Arc<Mutex<Vec<Bar>>> = d(),
    baz: i32 = 5,
}

let state = State { .. };
```

It *is* more verbose than absolutely needed, but it has some benefits as well.

`#[derive(const Default)]` is an expansion of what `derive` can do:

   - how do we teach it?
   - how googleable is it? how will people that don't know about this find out about it in the first place or find the appropriate documentation when they encounter it in the wild?
   - how clearly does it communicate what it does? Because it can be read as affecting the derived `impl Default` to be `const`, not changing the behavior of default values
   - if that is allowed, then does `#[derive(const Clone)]` work? what about `#[derive(const Trait)]`?

My biggest concern about this is that it introduces an entire new concept, that is an exception to other rules, for the benefit or removing 1 call * # of fields at worst. That's of course also setting aside that [we can't construct an `Arc` in `const`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e321d81465a207aed95195a80adf0a1e) today, but that's orthogonal.

Reply:
I'd like to see what @tmandry mentions added as a section in the alternatives (had the same thought while reading).

It just seems a bit non-symmetric that the specified defaults are used when implementing `Default` but the `Default` impls on fields aren't used with the `{ .. }` syntax, and this then creates the odd discrepancy between the meaning of `{ .. }` and `{ ..<_>::default() }`.

CC @yoshuawuyts @oli-obk @fmease, regarding the ongoing work to allow for `const` impls of `Default`.

Reply:
I would say this is more of a future possibility than an alternative. I should have explained I'm not saying we should do this for only `#[derive(Default)]`, but that we will most likely want a general `const` trait modifier, and probably several kinds of `#[derive(const ...)]`, and this would jive well with that. So it _would_ be changing the `impl Default` to be `const`.

Reply:
I do it as pretty important to eventually unify these two features, though.